Add logic that get all parsed dependencies without online

t0rchwo0d opened the pull request:
For overall dependency management in project, I Added `--parse-only` parse only command line switch that output parsed dependecies in lockfiles without vunerability search request.

Please feel free to let me know if you need any improvement for merge.


**Here's what the outputs :**

`.\osv-scanner.exe --parse-only -L .\fixtures\locks-many\composer.lock -L .\fixtures\locks-many\yarn.lock`

```json
{
  "results": [
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many\\yarn.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "balanced-match",
            "version": "1.0.2",
            "ecosystem": "npm"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many\\composer.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "sentry/sdk",
            "version": "2.0.4",
            "ecosystem": "Packagist"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    }
  ],
  "packages": null
}
```

`.\osv-scanner.exe --parse-only -r .\fixtures\`

```json
{
  "results": [
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many\\composer.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "sentry/sdk",
            "version": "2.0.4",
            "ecosystem": "Packagist"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many\\yarn.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "balanced-match",
            "version": "1.0.2",
            "ecosystem": "npm"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many-with-invalid\\Gemfile.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "ast",
            "version": "2.4.2",
            "ecosystem": "RubyGems"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many-with-invalid\\yarn.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "balanced-match",
            "version": "1.0.2",
            "ecosystem": "npm"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-one-with-nested\\nested\\composer.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "sentry/sdk",
            "version": "2.0.4",
            "ecosystem": "Packagist"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-one-with-nested\\yarn.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "balanced-match",
            "version": "1.0.2",
            "ecosystem": "npm"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    },
    {
      "source": {
        "path": "D:\\Project\\osv-scanner\\cmd\\osv-scanner\\fixtures\\locks-many\\Gemfile.lock",
        "type": "lockfile"
      },
      "packages": [
        {
          "package": {
            "name": "ast",
            "version": "2.4.2",
            "ecosystem": "RubyGems"
          },
          "vulnerabilities": null,
          "groups": []
        }
      ]
    }
  ],
  "packages": null
}
```

G-Rath reviewed (changes requested):
Thanks, this was on my todo list as something to bring across from [the `osv-detector`](https://github.com/G-Rath/osv-detector#--list-packages), but I think we shouldn't limit it to just JSON

Can you also make sure to add some tests, especially to the CLI tests?

t0rchwo0d commented:
> Thanks, this was on my todo list as something to bring across from [the `osv-detector`](https://github.com/G-Rath/osv-detector#--list-packages), but I think we shouldn't limit it to just JSON
> 
> Can you also make sure to add some tests, especially to the CLI tests?

Okay, ASAP I will add markdown & table output format.
And also add CLI tests case to "main_test.go".

If I miss understood "CLI tests", please let me know. : )
Thanks for your confirmed.

G-Rath commented:
> And also add CLI tests case to "main_test.go".
> If I miss understood "CLI tests", please let me know. : )

Nope, `main_test.go` was what I was meaning :)

t0rchwo0d commented:
Hi, @G-Rath

 I added following :
- Added logic for support output format (json, table, markdown)
- Updated main_test.go

I'd appreciate it if you could review that!

oliverchang commented:
Hi! Thank you for this contribution.

May we ask also what the intended use case of this is? Are you trying to generate something similar to an SBOM? This may help us in understanding how we can provide the best experience for your use case.

t0rchwo0d commented:
> Hi! Thank you for this contribution.
> 
> May we ask also what the intended use case of this is? Are you trying to generate something similar to an SBOM? This may help us in understanding how we can provide the best experience for your use case.

Hi, I thinking of building a very simple web dashboard to manage dependencies for security engineer in our company.
So, I needed json data, so I became interested osv-scanner and contributed to the project!  (Basically, I like Go)

I think, It would be really cool if SBOM generation function with license information and more was added in the future.

t0rchwo0d commented:
I think the last update before the review.
Please let me know if you have any requirements after the review! : )

t0rchwo0d commented:
I moved added function to exsiting files.

oliverchang commented:
Thanks for working on this PR @t0rchwo0d ! 

As is this seems like a bit of a shoehorned way of achieving what you want. Given our desire to keep our tool minimal (in terms of flags, configuration flags), and avoid breaking compatibility in the future, could we instead spend a bit more time thinking about how we can solve your use case a bit more cleanly? I've filed #207 to track this FR specifically.

t0rchwo0d commented:
Hi, @oliverchang
Thank you for the review!

I agree your opinion that all problems are clean if supporting SBOM-like format.

I think developing a structure in SBOM format is not complicated,
However we should be considering SBOM format (SPDX or CycloneDX or ...).

If I want to using the [Owasp Dependency-Track](https://dependencytrack.org/), I think decide to [CycloneDX SBOM format](https://cyclonedx.org/use-cases/). 

```json
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
  "version": 1,
  "components": [
    {
      "type": "library",
      "name": "acme-library",
      "version": "1.0.0"
    }
  ]
}
```

Come to think of it, there might be a way to support one first and then add another format.

I would appreciate it if you could tell me your opinion. 
If positive opinion, I will be modify output to SBOM format.

t0rchwo0d commented:
> Hi, @oliverchang Thank you for the review!
> 
> I agree your opinion that all problems are clean if supporting SBOM-like format.
> 
> I think developing a structure in SBOM format is not complicated, However we should be considering SBOM format (SPDX or CycloneDX or ...).
> 
> If I want to using the [Owasp Dependency-Track](https://dependencytrack.org/), I think decide to [CycloneDX SBOM format](https://cyclonedx.org/use-cases/).
> 
> ```json
> {
>   "bomFormat": "CycloneDX",
>   "specVersion": "1.4",
>   "serialNumber": "urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79",
>   "version": 1,
>   "components": [
>     {
>       "type": "library",
>       "name": "acme-library",
>       "version": "1.0.0"
>     }
>   ]
> }
> ```
> 
> Come to think of it, there might be a way to support one first and then add another format.
> 
> I would appreciate it if you could tell me your opinion. If positive opinion, I will be modify output to SBOM format.

Already used "github.com/CycloneDX/cyclonedx-go" in project, so I can use that.

t0rchwo0d commented:
Hi, @oliverchang 
I added logic to output in SBOM format of CycloneDX.
This is experimental feature feeling to me.

Anyway, I tested about mvn, npm, pypi, go.
I hope this will help the project : )

**Use Case**
`output SBOM` -> `put SBOM in osv-scanner` -> `check results`

**Result**
`./osv-scanner -f sbom -parse-only ./requirements.txt`
```json
Scanning dir ./requirements.txt
Scanned /mnt/c/Project/osv-scanner/cmd/osv-scanner/requirements.txt file and found 4 packages
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "version": 1,
  "components": [
    {
      "bom-ref": "pkg:pypi/aiohttp@3.5.3",
      "type": "library",
      "name": "aiohttp",
      "version": "3.5.3",
      "purl": "pkg:pypi/aiohttp@3.5.3"
    },
    {
      "bom-ref": "pkg:pypi/aiohttp-jinja2@1.1.0",
      "type": "library",
      "name": "aiohttp-jinja2",
      "version": "1.1.0",
      "purl": "pkg:pypi/aiohttp-jinja2@1.1.0"
    },
    {
      "bom-ref": "pkg:pypi/aiohttp-session@2.7.0",
      "type": "library",
      "name": "aiohttp-session",
      "version": "2.7.0",
      "purl": "pkg:pypi/aiohttp-session@2.7.0"
    },
    {
      "bom-ref": "pkg:pypi/aiopg@0.15.0",
      "type": "library",
      "name": "aiopg",
      "version": "0.15.0",
      "purl": "pkg:pypi/aiopg@0.15.0"
    }
  ]
}
```

`./osv-scanner ./test.json`
```shell
Scanning dir ./test.json
Scanned CycloneDX SBOM and found 4 packages
╭─────────────────────────────────────┬───────────┬─────────┬─────────┬───────────╮
│ OSV URL (ID IN BOLD)                │ ECOSYSTEM │ PACKAGE │ VERSION │ SOURCE    │
├─────────────────────────────────────┼───────────┼─────────┼─────────┼───────────┤
│ https://osv.dev/GHSA-v6wp-4m6f-gcjg │ PyPI      │ aiohttp │ 3.5.3   │ test.json │
│ https://osv.dev/PYSEC-2021-76       │           │         │         │           │
╰─────────────────────────────────────┴───────────┴─────────┴─────────┴───────────╯
```

oliverchang commented:
hey @t0rchwo0d, sorry for the slow response on this PR.

As we commit to long-term supporting OSV-Scanner and all its features, we would like to spend a bit more time making sure we're making the right usability/maintenance decisions with our tool.

So far it seems like there are many other focused tools available that are capable of generating SBOMs, and it wasn't clear if we'd be adding much to this ecosystem if we support this inside OSV-Scanner. If this changes, we'll be sure to let you know :)
